### PR TITLE
Add multiple models

### PR DESCRIPTION
**Pull request by gyfbianhuanyun**
1. Add model options:
a. RNN (with bidirection or not)
b. LSTM (with bidirection or not)
c. GRU (with bidirection or not)
d. FC (with time points)
2. Add brain region options:
a. Left (the region name with left in ALL2)
b. Right (the region name with right in ALL2)
c. All regions

### CODE REVIEW

**Comment by gyfbianhuanyun** [comment]
@albert-no 
교수님, 수정했습니다.
이제 여러 model과 상황을 따라서 training 할 수 있습니다~

File: brain_all_FC.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        self.num_layers = layers
+
+        self.layer1 = nn.Sequential(nn.Linear(input_dim, hidden_dim),
+                                    nn.ReLU(True),
+                                    nn.Dropout(drop_prob))
+        self.hidden = nn.ModuleList()
+        for i in range(1, self.num_layers):
+            self.hidden.append(nn.Linear(hidden_dim, hidden_dim))
+            self.hidden.append(nn.ReLU(True))
+            self.hidden.append(nn.Dropout(drop_prob))
```
Comment (albert-no) [inline brain_all_FC.py:18]: 여기도 batch norm을 추가하겠습니다.

File: brain_all_FC.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        self.fc1 = nn.Linear(hidden_dim, 1, bias=True)
+        self.relu = nn.ReLU(inplace=True)
+        self.dropout = nn.Dropout(p=drop_prob)
+        self.fc2 = nn.Linear(100, output_dim, bias=True)
+
+        self.init_weights()
+
+        # weight_init
```
Comment (albert-no) [inline brain_all_FC.py:26]: comment지우고 flake8도 한번더 확인해보기 바랍니다.

File: brain_all_FC.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+def get_tensor_fc(device, data_x, data_y, start_idx, end_idx):
+    x_tensor = torch.FloatTensor(data_x[start_idx:end_idx, :, :]).to(device)
+    y_tensor = torch.FloatTensor(data_y[start_idx:end_idx])
+    y_tensor.unsqueeze_(-1)
+    y_tensor = y_tensor.to(device)
+    return x_tensor, y_tensor
+
+
+# Use GRU or CPU
+def get_device():
```
Comment (albert-no) [inline brain_all_FC.py:58]: utils.py나 misc.py등을 만들어서 잡다한 함수들을 모아두기 바랍니다.
모델을 정의하는 python파일과 train및 validation함수를 정의하는 python도 분리하는게 좋겠습니다.

File: brain_utils.py [hunk 4]
```diff
@@ -35,14 +36,25 @@ def get_data(param):
...
 
         subject_data = np.genfromtxt(subject_data_path, delimiter=',')
-        subject_x = torch.FloatTensor((subject_data[1:-1, :region_n] - minmax_x[0]) / (minmax_x[1] - minmax_x[0]))
+        if brain_region == 'all':
+            subject_idx = np.arange(0, 94, 1)
+        elif brain_region == 'left':
+            subject_idx = np.arange(0, 94, 2)
+        elif brain_region == 'right':
+            subject_idx = np.arange(1, 94, 2)
+        if model_name == 'FC':
```
Comment (albert-no) [inline brain_utils.py:45]: modelname에 따라 data를 불러오는 함수를 utils.py등에 정의하기 바랍니다.

File: brain_utils.py [hunk 5]
```diff
@@ -35,14 +36,25 @@ def get_data(param):
...
+            subject_x = torch.FloatTensor((subject_data[1:-1, subject_idx] - minmax_x[0]) / (minmax_x[1] - minmax_x[0]))
         subject_y = row.new_age
 
         data_x.append(subject_x)
         data_y.append(subject_y)
         length.append(subject_x.shape[0])
 
     # Normalize data
+    if model_name == 'FC':
+        data_x = (np.array(data_x) - minmax_x[0]) / (minmax_x[1] - minmax_x[0])
```
Comment (albert-no) [inline brain_utils.py:57]: normalize함수를 잘정의해서 여기서도 재활용이 가능하면 좋겠네요.

File: brain_wrapper.py [hunk 6]
```diff
@@ -51,8 +55,15 @@ def wrapper(param, data_x, data_y, length_x, learning_rate, lr_gamma, hidden_dim
     step_list = []
     epoch_list = []
 
-    mynet = RNNClassifier(
-        input_dim, hidden_dim, output_dim, layers, drop_prob).to(device)
+    if brain == 'right' or brain == 'left':
+        input_dim = input_dim // 2
+
+    if model_name == 'FC':
```
Comment (albert-no) [inline brain_wrapper.py:61]: mynet을 불러오는 함수를 정의하기 바랍니다.
if문이 코드안에 너무 많으면 읽기가 어렵습니다.

File: brain_wrapper.py [hunk 7]
```diff
@@ -74,11 +85,15 @@ def wrapper(param, data_x, data_y, length_x, learning_rate, lr_gamma, hidden_dim
         loss = 0
 
         for tr in range(int(train_num / rate_tr)):
-            train_x_tensor, train_y_tensor, train_length_tensor = train(
-                device, tr*rate_tr, rate_tr, train_xdata, train_ydata, train_length_x)
+            if model_name == 'FC':
```
Comment (albert-no) [inline brain_wrapper.py:88]: 역시 model_name에 따라 다른방식으로 학습시키는 train함수를 만들기 바랍니다.

File: brain_wrapper.py [hunk 8]
```diff
@@ -96,10 +111,14 @@ def wrapper(param, data_x, data_y, length_x, learning_rate, lr_gamma, hidden_dim
         valid_loss = 0
 
         for va in range(int(valid_num / rate_va)):
-            valid_x_tensor, valid_y_tensor, valid_length_tensor = valid(
-                device, va*rate_va, rate_va, valid_xdata, valid_ydata, valid_length_x)
-            valid_x_tensor = pack_padded_sequence(valid_x_tensor, valid_length_tensor,
-                                                  batch_first=True, enforce_sorted=False)
+            if model_name == 'FC':
```
Comment (albert-no) [inline brain_wrapper.py:114]: valid함수도 model_name에 따라 다르게 validation하는 함수로 만들기 바랍니다.

File: brain_wrapper.py [hunk 9]
```diff
@@ -133,10 +152,14 @@ def wrapper(param, data_x, data_y, length_x, learning_rate, lr_gamma, hidden_dim
 
     mynet.eval()
     with torch.no_grad():
-        test_x_tensor, test_y_tensor, test_length_tensor = get_tensor(
-            device, data_x, data_y, length_x, train_num + valid_num, total_num)
-        test_x_tensor = pack_padded_sequence(test_x_tensor, test_length_tensor,
-                                             batch_first=True, enforce_sorted=False)
+        if model_name == 'FC':
```
Comment (albert-no) [inline brain_wrapper.py:155]: get_tensor도 if문없이 만들어보기바랍니다.

**Review by albert-no — COMMENTED** [review]
코드를 수정하는김에 조금 다듬으면 좋겠습니다.
함수들을 파일별로 잘 분류해보기 바랍니다.

**Comment by gyfbianhuanyun** [comment]
수정했습니다~

File: brain_utils.py [hunk 10]
```diff
@@ -13,6 +13,41 @@ def get_data_path(data_folder):
...
+        data_x = data[1:time+1, idx]
+    else:
+        data_x = torch.FloatTensor((data[1:-1, idx] - minmax_x[0]) /
+                                   (minmax_x[1] - minmax_x[0]))
+
+    return data_x
+
+
+# Normalize data
+def normaize_data(datax, datay, model_name, minmax_x, minmax_y):
```
Comment (albert-no) [inline brain_utils.py:42]: normalize

File: brain_utils.py [hunk 11]
```diff
@@ -35,23 +71,27 @@ def get_data(param):
...
+                                       subject_idx, minmax_x)
         subject_y = row.new_age
 
         data_x.append(subject_x)
         data_y.append(subject_y)
         length.append(subject_x.shape[0])
 
-    # Normalize data
-    data_y = (np.array(data_y) - minmax_y[0]) / (minmax_y[1] - minmax_y[0])
+    data_x, data_y = normaize_data(data_x, data_y, model_name,
```
Comment (albert-no) [inline brain_utils.py:84]: normalize

File: comparision.py [hunk 12]
```diff
@@ -6,43 +6,55 @@
...
 
+
 def file_name(file_dir):
     filelist = []
     for file in os.listdir(file_dir):
         filelist.append(file)
     return filelist
 
+
 def comparision(project_name, file_name):
```
Comment (albert-no) [inline comparision.py:17]: comparison
파일이름도 바뀌어야 할것같네요.

**Review by albert-no — COMMENTED** [review]
네 수정하는 부분이 너무 늘어나고있는것같네요.
몇가지 수정하고 merge하도록 하겠습니다.
각 model별로 (parameter별로) excel파일을 만들어 결과를 정리해보기 바랍니다.

**Comment by gyfbianhuanyun** [comment]
수정했습니다.
`comparision.py` 이름은` load_data.py` 로 바꿨습니다.
그리고 안에 함수 이름도 `get_reads_through_comparision` 로 바꿨습니다.

**Comment by albert-no** [comment]
comparision대신 comparison을 사용하기 바랍니다.

**Comment by gyfbianhuanyun** [comment]
네, 알겠습니다~

**Comment by albert-no** [comment]
comparison부분만 수정하여 merge하기 바랍니다.
